klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
         self.gcode = self.printer.lookup_object('gcode')
         self.gcode.register_command("BLTOUCH_DEBUG", self.cmd_BLTOUCH_DEBUG,
                                     desc=self.cmd_BLTOUCH_DEBUG_help)
+        # Debugging code to report sensor pin state changes
+        ppins.reset_pin_sharing(pin_params)
+        buttons = self.printer.try_load_module(config, "buttons")
+        buttons.register_buttons([pin], self.sensor_debug_callback)
+    def sensor_debug_callback(self, eventtime, state):
+        logging.info("bltouch sensor @ %.6f is %s", eventtime, state)
     def _build_config(self):
         kin = self.printer.lookup_object('toolhead').get_kinematics()
         for stepper in kin.get_steppers('Z'):
             stepper.add_to_endstop(self)
-    def send_cmd(self, print_time, cmd):
-        self.mcu_pwm.set_pwm(print_time, Commands[cmd] / SIGNAL_PERIOD)
-    def test_sensor(self):
-        if not self.test_sensor_pin:
-            return
+    def handle_connect(self):
+        self.sync_mcu_print_time()
+        try:
+            self.raise_probe()
+        except homing.EndstopError as e:
+            raise self.printer.config_error(str(e))
+    def sync_mcu_print_time(self):
+        curtime = self.printer.get_reactor().monotonic()
+        est_time = self.mcu_pwm.get_mcu().estimated_print_time(curtime)
+        self.next_cmd_time = max(self.next_cmd_time, est_time + MIN_CMD_TIME)
+    def sync_print_time(self):
         toolhead = self.printer.lookup_object('toolhead')
         print_time = toolhead.get_last_move_time()
-        if print_time < self.next_test_time:
-            self.next_test_time = print_time + TEST_TIME
-            return
-        # Raise the bltouch probe and test if probe is raised
-        self.send_cmd(print_time, 'reset')
-        home_time = print_time + self.pin_move_time
-        self.send_cmd(home_time, 'touch_mode')
-        self.send_cmd(home_time + MIN_CMD_TIME, None)
-        # Perform endstop check to verify bltouch reports probe raised
+        if self.next_cmd_time > print_time:
+            toolhead.dwell(self.next_cmd_time - print_time)
+        else:
+            self.next_cmd_time = print_time
+    def send_cmd(self, cmd, duration=MIN_CMD_TIME):
+        self.mcu_pwm.set_pwm(self.next_cmd_time, Commands[cmd] / SIGNAL_PERIOD)
+        # Translate duration to ticks to avoid any secondary mcu clock skew
+        mcu = self.mcu_pwm.get_mcu()
+        cmd_clock = mcu.print_time_to_clock(self.next_cmd_time)
+        cmd_clock += mcu.seconds_to_clock(max(duration, MIN_CMD_TIME))
+        self.next_cmd_time = mcu.clock_to_print_time(cmd_clock)
+        return self.next_cmd_time
+    def verify_state(self, check_start_time, check_end_time, triggered, msg):
```

Comment:
I don't understand what you are reporting here.

Reply:
I'm reporting that the state of the pin can go like this:
`CMD` -> pin 0 -> "time it takes to detect the error" -> pin 1

For simplicity of the configuration, the time it takes to detect the error is built in in the `pin_move_time`

So, we need to check that the pin is still zero after `pin_move_time` has expired. But the code snippet above sets `self._min_query_time = self._mcu.monotonic()`, which means that it starts to check for the zero state immediately, rather than after `print_time`, print time is only used for commanding the movement.

I could be wrong, but this is my understanding of the code, and also my practical experience when I tried to do something like you did there. It was simply reading too early and reporting that everything was good, when it in fact wasn't.

Reply:
Well, to me it looks like `min_query_time` is the only variable that ensures that the `_check_busy` function does not check the homing state until a given time (to ensure that we are not reading some state from previous operations).  

```python
    def _check_busy(self, eventtime, home_end_time=0.):
        # Check if need to send an end_stop_query command
        last_sent_time = self._last_state.get('#sent_time', -1.)
        if last_sent_time >= self._min_query_time or self._mcu.is_fileoutput():
            if not self._homing:
                return False
            if not self._last_state.get('homing', 0):
                for s in self._steppers:
                    s.note_homing_end(did_trigger=True)
                self._homing = False
                return False
            last_sent_print_time = self._mcu.estimated_print_time(
                last_sent_time)
            if last_sent_print_time > home_end_time:
                # Timeout - disable endstop checking
                for s in self._steppers:
                    s.note_homing_end()
                self._homing = False
                self._home_cmd.send([self._oid, 0, 0, 0, 0, 0])
                raise self.TimeoutError("Timeout during endstop homing")
        if self._mcu.is_shutdown():
            raise error("MCU is shutdown")
        est_print_time = self._mcu.estimated_print_time(eventtime)
        if est_print_time >= self._next_query_print_time:
            self._next_query_print_time = est_print_time + self.RETRY_QUERY
            self._query_cmd.send([self._oid])
        return True
```
But I have to admit, I did not read the homing code that carefully, because I was using the `query_endstop` / `query_endstop_wait` pair of functions, which are broken in a similar way

```python
    def query_endstop(self, print_time):
        self._homing = False
        self._min_query_time = self._mcu.monotonic()
        self._next_query_print_time = print_time
    def query_endstop_wait(self):
        eventtime = self._mcu.monotonic()
        while self._check_busy(eventtime):
            eventtime = self._mcu.pause(eventtime + 0.1)
        return self._last_state.get('pin', self._invert) ^ self._invert
```

The problem there, is that even if the `_next query_print_time` is set. the loop will end prematurely, and we read the previous pin state. 

I'm also curious why you are using fake homing operations, rather than quering the endstops directly, since the former seems more of a hack to get some very specific behaviour? Maybe because of the EMI concerns you raised in another comment?